File: 1074_num_submatrix_sum_target/num_submatrix_sum_target.go
/*
 * @Date: 2021-05-29 09:38:37
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-29 09:54:30
 */
package main

func numSubmatrixSumTarget(matrix [][]int, target int) (ans int) {
	subarraySum := func(nums []int, k int) (ans int) {
		mp := map[int]int{0: 1}
		for i, pre := 0, 0; i < len(nums); i++ {
			pre += nums[i]
			if _, ok := mp[pre-k]; ok {
				ans += mp[pre-k]
			}
			mp[pre]++
		}
		return
	}
	for i := range matrix { // 枚举上边界
		sum := make([]int, len(matrix[0]))
		for _, row := range matrix[i:] { // 枚举下边界
			for c, v := range row {
				sum[c] += v // 更新每列的元素和
			}
			ans += subarraySum(sum, target)
		}
	}
	return
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed!")
		}
	}

	assert(numSubmatrixSumTarget([][]int{{0, 1, 0}, {1, 1, 1}, {0, 1, 0}}, 0) == 4)
	assert(numSubmatrixSumTarget([][]int{{1, -1}, {-1, 1}}, 0) == 5)
	assert(numSubmatrixSumTarget([][]int{{904}}, 0) == 0)
}

File: 1074_num_submatrix_sum_target/num_submatrix_sum_target.rs
/*
 * @Date: 2021-05-29 09:38:33
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-05-29 10:13:43
 */

fn num_submatrix_sum_target(mut matrix: Vec<Vec<i32>>, target: i32) -> i32 {
    use std::collections::HashMap;
    let m = matrix.len();
    let n = matrix[0].len();
    let mut ans = 0;
    //前缀预处理
    for i in 0..m {
        for j in 1..n {
            matrix[i][j] += matrix[i][j - 1];
        }
    }

    for k in 0..n {
        for j in k..n {
            let mut sum = 0;
            let mut mp: HashMap<i32, i32> = HashMap::new();
            mp.insert(0, 1);
            for i in 0..m {
                if k == 0 {
                    sum += matrix[i][j]
                } else {
                    sum += matrix[i][j] - matrix[i][k - 1];
                }
                if mp.contains_key(&(sum - target)) {
                    ans += *(mp.entry(sum - target).or_insert(0));
                }
                let count = mp.entry(sum).or_insert(0);
                *count += 1;
            }
        }
    }
    ans
}

fn main() {
    assert_eq!(
        num_submatrix_sum_target(vec![vec![0, 1, 0], vec![1, 1, 1], vec![0, 1, 0]], 0),
        4
    );
    assert_eq!(
        num_submatrix_sum_target(vec![vec![-1, 1], vec![1, -1]], 0),
        5
    );
    assert_eq!(num_submatrix_sum_target(vec![vec![904]], 0), 0);
}
